### PR TITLE
Sketch alternative multicast model – #1021 reloaded

### PR DESCRIPTION
**Pull request by philsbln**
This is a remake of #1021 as @GrumpyOldTroll suggested.
It moves the WithIPvXMulticastGroup for receiving to the LocalEndpoint and adds the ASM/SSM distinguisher.
It does not add TTL support.

### CODE REVIEW

**Comment by philsbln** [comment]
I got more in-person input from @GrumpyOldTroll in person in Philly.
We might want to include the src addr for SSN into a "WithSSNMulticastGroup" endpoint attribute and make Multicast addresses "special" that way. 

I guess we don't want to support racing multicast against non-multicast for now.

**Review by MaxF12 — APPROVED** [review]
LGTM, one minor nit

**Review by britram — APPROVED** [review]
LGTM but I'd like @csperkins eyes on this as well.

**Review by GrumpyOldTroll — COMMENTED** [review]
Overall I think this looks like it should be fine.

The things that are not currently clear to me (might be covered elsewhere already? I haven't re-reviewed the whole doc):
- What if any difference is there between a sender using RemoteSpecifier.WithMulticastGroupIPvX and RemoteSpecifier.WithIPvXAddress?  Would WithIPvXAddress raise an error if given a group address, or would it use different defaults for the packets, or what exactly? (I'm thinking here of TTL/hoplimit and destination mac address as the 2 examples that come to mind that could have different defaults for these cases.)
- how is the default interface determined if not provided explicitly?  (sockets with apis that allow it I think will use the route to the group address, but I think ssm receive probably should use the route to the source address)
- will the listener merge receive data from different source addresses and source ports as packets are received into the same connection object, or will each new source (addr,port) pair result in a new connection, or is there some kind of controller that can pick between these?

Outside of these, I haven't noticed any questions without answers that seem obvious to me.

File: draft-ietf-taps-interface.md [hunk 1]
```diff
@@ -737,23 +737,60 @@ early binding when required, for example with some Network Address Translator
...
+RemoteSpecifier.WithMulticastGroupIPv6(GroupAddress)
+RemoteSpecifier.WithPort(PortNumber)
+```
+
+Calling Listen() on a Preconnection with a multicast group specified on the Remote
+Endpoint will join the multicast group to receive messages. This Listener
+will create one Connection for each Remote Endpoint sending to the group,
+with the Local Endpoint set to the group address. The set of Connection
+objects created forms a Connection Group.
+
```
Comment (csperkins) [inline draft-ietf-taps-interface.md:762]: I think this makes sense, but I wonder if the Local Endpoint must somehow specify both the group address and the local address/port to which the connection is bound? 

Or, to put this another way, when calling `Listen()` is it a requirement to bind to a specific local interface? If not, does the system pick a default interface from which to join the group, or does it join the group from all available interfaces? If the former, how can the application know what interface was used? If the latter, how can it tell on what interface each packet was received?

File: draft-ietf-taps-interface.md [hunk 2]
```diff
@@ -737,23 +737,60 @@ early binding when required, for example with some Network Address Translator
...
+way are send-only, and do not join the multicast group. The resulting
+Connection will have a Local Endpoint indicating the local interface to
+which the connection is bound and a Remote Endpoint indicating the
+multicast group.
+
+```
+RemoteSpecifier.WithMulticastGroupIPv4(GroupAddress)
+RemoteSpecifier.WithMulticastGroupIPv6(GroupAddress)
+RemoteSpecifier.WithPort(PortNumber)
+```
```
Comment (csperkins) [inline draft-ietf-taps-interface.md:755]: RemoteSpecifier.WithTTL(TTL)

File: draft-ietf-taps-interface.md [hunk 3]
```diff
@@ -737,23 +737,60 @@ early binding when required, for example with some Network Address Translator
...
+group address as the Local Endpoint results in an EstablishmentError.
+
+```
+RemoteSpecifier.WithMulticastGroupIPv4(GroupAddress)
+RemoteSpecifier.WithMulticastGroupIPv6(GroupAddress)
+RemoteSpecifier.WithPort(PortNumber)
+LocalSpecifier.WithAnySourceMulticastGroupIPv4(GroupAddress)
+LocalSpecifier.WithAnySourceMulticastGroupIPv6(GroupAddress)
+LocalSpecifier.WithPort(PortNumber)
+```
```
Comment (csperkins) [inline draft-ietf-taps-interface.md:789]: RemoteSpecifier.WithTTL(TTL)

**Comment by csperkins** [comment]
This will close #874 if merged.

File: draft-ietf-taps-interface.md (hunk 1 above)
Comment (philsbln) [reply draft-ietf-taps-interface.md:762]: I discussed this with @GrumpyOldTroll in Philly … and we decided to leave that unspecified. We assume the implementation will throw an error if it has insufficient information. In the examples, we alway give an Interface, but a source address might also be sufficient.

File: draft-ietf-taps-interface.md [hunk 4]
```diff
@@ -737,23 +737,64 @@ early binding when required, for example with some Network Address Translator
...
-that are accepted from this Listener are receive-only.
-
-Similarly, specifying a multicast group address on the Remote Endpoint will indicate that the
-resulting connection will be used to send multicast messages, and that the Preconnection will
-support Initiate() but not Listen(). Any Connections created this way are send-only.
-
-A Rendezvous() call on Preconnections containing group addresses results in an
-EstablishmentError as described in {{rendezvous}}.
+To use multicast, a Preconnection is first created with the Local/Remote Endpoint
+specifying the ASM/SSM multicast group and destination port number.
```
Comment (tfpauly) [inline draft-ietf-taps-interface.md:741]: We need to expand ASM/SSM the first time I think

File: draft-ietf-taps-interface.md [hunk 5]
```diff
@@ -737,23 +737,64 @@ early binding when required, for example with some Network Address Translator
...
+way are send-only, and do not join the multicast group. The resulting
+Connection will have a Local Endpoint indicating the local interface to
+which the connection is bound and a Remote Endpoint indicating the
+multicast group.
+
+```
+RemoteSpecifier.WithMulticastGroupIPv4(GroupAddress)
+RemoteSpecifier.WithMulticastGroupIPv6(GroupAddress)
+RemoteSpecifier.WithPort(PortNumber)
+RemoteSpecifier.WithTTL(TTL)
```
Comment (tfpauly) [inline draft-ietf-taps-interface.md:755]: Move this up before the prior paragraph?

Can we also describe the TTL parameter?

File: draft-ietf-taps-interface.md [hunk 6]
```diff
@@ -737,23 +737,64 @@ early binding when required, for example with some Network Address Translator
...
+LocalSpecifier.WithAnySourceMulticastGroupIPv4(GroupAddress)
+LocalSpecifier.WithAnySourceMulticastGroupIPv6(GroupAddress)
+LocalSpecifier.WithPort(PortNumber)
+LocalSpecifier.WithTTL(TTL)
+```
 
 See {{multicast-examples}} for more examples.
 
+
```
Comment (tfpauly) [inline draft-ietf-taps-interface.md:797]:
```suggestion
```

File: draft-ietf-taps-interface.md (hunk 1 above)
Comment (philsbln) [reply draft-ietf-taps-interface.md:763]:
```suggestion
The receiving interface can be restricted by passing it as part of the LocalSpecifier or queried through the MessagContext on the messages received (see {{msg-ctx}} for further details).

```

File: draft-ietf-taps-interface.md [hunk 7]
```diff
@@ -737,20 +737,60 @@ early binding when required, for example with some Network Address Translator
...
-that are accepted from this Listener are receive-only.
-
-Similarly, specifying a multicast group address on the Remote Endpoint will indicate that the
-resulting connection will be used to send multicast messages, and that the Preconnection will
-support Initiate() but not Listen(). Any Connections created this way are send-only.
-
-A Rendezvous() call on Preconnections containing group addresses results in an
-EstablishmentError as described in {{rendezvous}}.
+To use multicast, a Preconnection is first created with the Local/Remote Endpoint
+specifying the ASM/SSM multicast group and destination port number.
```
[the remark this answers is not in the document]
Comment (tfpauly) [reply draft-ietf-taps-interface.md:741]:
```suggestion
specifying the Any-source multicast (ASM) or Source-specific multicast (SSM) multicast group and destination port number.
```

File: draft-ietf-taps-interface.md [hunk 8]
```diff
@@ -737,20 +737,60 @@ early binding when required, for example with some Network Address Translator
...
-that are accepted from this Listener are receive-only.
-
-Similarly, specifying a multicast group address on the Remote Endpoint will indicate that the
-resulting connection will be used to send multicast messages, and that the Preconnection will
-support Initiate() but not Listen(). Any Connections created this way are send-only.
-
-A Rendezvous() call on Preconnections containing group addresses results in an
-EstablishmentError as described in {{rendezvous}}.
+To use multicast, a Preconnection is first created with the Local/Remote Endpoint
+specifying the _any source multicast (ASM)_ / _source specific multicast (SSM)_ multicast group and destination port number.
```
Comment (tfpauly) [inline draft-ietf-taps-interface.md:741]:
```suggestion
specifying the any-source multicast (ASM) or source-specific multicast (SSM) multicast group and destination port number.
```

File: draft-ietf-taps-interface.md [hunk 9]
```diff
@@ -737,20 +737,60 @@ early binding when required, for example with some Network Address Translator
...
+specifying the any-source multicast (ASM) or source-specific multicast (SSM) multicast group and destination port number.
+
+Calling Initiate() on that Preconnection creates a Connection that can be
+used to send messages to the multicast group. The Connection object that is
+created will support Send() but not Receive(). Any Connections created this
+way are send-only, and do not join the multicast group. The resulting
+Connection will have a Local Endpoint indicating the local interface to
+which the connection is bound and a Remote Endpoint indicating the
+multicast group.
+
```
Comment (csperkins) [inline draft-ietf-taps-interface.md:750]: add "The following API calls can be used to configure a `Preconnection` before calling `Initiate()`:

File: draft-ietf-taps-interface.md [hunk 10]
```diff
@@ -737,20 +737,60 @@ early binding when required, for example with some Network Address Translator
...
+To use multicast, a Preconnection is first created with the Local/Remote Endpoint
+specifying the any-source multicast (ASM) or source-specific multicast (SSM) multicast group and destination port number.
+
+Calling Initiate() on that Preconnection creates a Connection that can be
+used to send messages to the multicast group. The Connection object that is
+created will support Send() but not Receive(). Any Connections created this
+way are send-only, and do not join the multicast group. The resulting
+Connection will have a Local Endpoint indicating the local interface to
+which the connection is bound and a Remote Endpoint indicating the
+multicast group.
```
Comment (mwelzl) [inline draft-ietf-taps-interface.md:749]:
```suggestion
multicast group.

The following API calls can be used to configure a Preconnection before calling Initiate():
```

**Review by mwelzl — APPROVED** [review]

**Pull request merged**